### DISCUSSION TITLE
Overriding the `sx` prop

### DISCUSSION THREAD

**Discussion by midrizi** [nandorojo/dripsy#281] [Q&A]
@nandorojo 

Did something change in terms of overriding `sx` prop in `v4`? 

Here's one of the Components:
```tsx
import { ComponentProps, FC } from 'react'
import { StyleSheet } from 'react-native'

import { View as DripView } from 'dripsy'
import { observer } from 'mobx-react-lite'

export type DividerProps = Pick<ComponentProps<typeof DripView>, 'sx'>

export const Divider: FC<DividerProps> = observer(function Divider({ sx, ...rest }) {
  return (
    <DripView
      sx={{
        bg: '$greyscale.border',
        height: () => StyleSheet.hairlineWidth,
        ...sx, // ERROR THROWN HERE WHEN OVERRIDING
      }}
      {...rest}
    />
  )
})

```

And the error I'm receiving:
```ts

Type '{ transform?: (PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]; ... 8 more ...; height: () => number; } | { ...; }' is not assignable to type 'SxProp'.
  Type '{ transform?: (PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]; ... 8 more ...; height: () => number; }' is not assignable to type 'SxProp'.
    Type '{ transform?: (PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]; ... 8 more ...; height: () => number; }' is not assignable to type 'Sx'.
      Type '{ transform?: (PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]; ... 8 more ...; height: () => number; }' is not assignable to type '{ [x: string]: FactoryValue<ResponsiveValue<number | "root" | (string & {}) | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathO...'.
        Property 'transform' is incompatible with index signature.
          Type '(PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]' is not assignable to type 'FactoryValue<ResponsiveValue<number | "root" | (string & {}) | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...> | D...'.
            Type '(PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform)[]' is not assignable to type '(number | "root" | (string & {}) | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...>)[]'.
              Type 'PerpectiveTransform | RotateTransform | RotateXTransform | RotateYTransform | RotateZTransform | ... 7 more ... | MatrixTransform' is not assignable to type 'number | "root" | (string & {}) | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...>'.
                Type 'PerpectiveTransform' is not assignable to type 'number | "root" | (string & {}) | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...>'.ts(2322)
sx.d.ts(59, 5): The expected type comes from property 'sx' which is declared here on type 'IntrinsicAttributes & Omit<PropsWithChildren<ViewProps>, "variant" | "variants"> & StyledProps<undefined> & RefAttributes<...>'

```

**Comment by nandorojo** [comment]
That particular error is odd, but yeah you can't necessarily spread it since it could also be a function. What if, rather than `Pick`, you import `Sx` from dripsy as the type?

**Reply by midrizi** [reply]
I did try `Sx` and the `SxProp`, same issue

![image](https://user-images.githubusercontent.com/56638730/231611171-fb47e6cb-74da-44fc-bb5f-c946a44014b9.png)

**Reply by nandorojo** [reply]
Weird, it’s working fine for me. Are you on an outdated `@types/react-native` package? Try 0.71.3

**Reply by midrizi** [reply]
I didn't have it since types are bundled as of `react-native@0.71`

https://reactnative.dev/blog/2023/01/12/version-071#typescript-by-default

> We’re also offering built-in, more accurate TypeScript declarations directly from the react-native package. This means you won’t need @types/react-native any longer, and the types will be updated in lockstep with React Native releases.

Anyways I installed it but still the same error.

**Reply by nandorojo** [reply]
Weird, I’m not able to reproduce it on my end

**Reply by nandorojo** [reply]
Maybe it’s something related to your theme

**Reply by midrizi** [reply]
Very weird indeed. 

Reset the theme to bare minimum and still the same issue
```ts
const baseTheme = makeTheme({
  types: {
    onlyAllowThemeValues: 'always',
  },
})

export type MyTheme = typeof baseTheme
declare module 'dripsy' {
  interface DripsyCustomTheme extends MyTheme {}
}
```

**Reply by nandorojo** [reply]
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZQB4Bo4gIYGsCmAKgBb4j5wC+cAZlBCHAOQAmUwYAzgJ5MBQfAMYQAdp3gxS5OAF4seIlPwAKBHzga4wgDbROALkTrNJgCQBzKPm6dB2bfkNqTLzQCNoLfFENM32gFd8flcNSmMwvkoASgEYbjAKEjIKOXjEiBo4SRSBL0FtbCssCBYAh2Y2Dh4mIxNgERhvGmxBCgARdi5uAGEA8QZk6XxUJpEWTjghigRw8L4HeE5UQzRZRHnFuGWAJlXUOWd3c18LKxs7BwA6DygvKCZ0CLhSYHNiGENlaNkAPjgAAxPExXUHLKJAA

Works fine here…

**Reply by midrizi** [reply]
Hmm... something else is going on here, you could be right about it being a `react-native` issue.

![image](https://user-images.githubusercontent.com/56638730/231792178-248e5436-f215-40bb-aec2-7f3f26faec15.png)


> Type '"center"' is not assignable to type 'FactoryValue<ResponsiveValue<"root" | DotPathOrKeyOf<{ readonly $0: 0; readonly $1: 4; readonly $2: 8; readonly $3: 12; readonly $4: 16; readonly $5: 24; readonly $6: 32; readonly $7: 56; readonly $8: 64; readonly $9: 128; readonly $10: 256; }> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...> | DotPathOrKeyOf<...>>>'.ts(2322)

**Reply by nandorojo** [reply]
what about `yarn why @types/react-native`

**Reply by midrizi** [reply]
I did upgrade `react-native@0.71.5` but no luck.

Also installed `@types/react-native@0.71.5` but I don't think it's using them, since types now are built in.

Here's the output:
```
yarn why v1.22.19
[1/4] Why do we have the module "@types/react-native"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@types/react-native@0.71.5"
info Reasons this module exists
   - "_project_#@root#native" depends on it
   - Hoisted from "_project_#@root#native#@types#react-native"
info Disk size without dependencies: "804KB"
info Disk size with unique dependencies: "1000KB"
info Disk size with transitive dependencies: "2.16MB"
info Number of shared dependencies: 2
Done in 1.88s.
```
And for `react-native`
```
[1/4] Why do we have the module "react-native"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "react-native@0.71.5"
info Reasons this module exists
   - "_project_#@root#native" depends on it
   - Hoisted from "_project_#@root#native#react-native"
info Disk size without dependencies: "82.43MB"
info Disk size with unique dependencies: "144.83MB"
info Disk size with transitive dependencies: "82.43MB"
info Number of shared dependencies: 386
Done in 2.44s
```

**Reply by nandorojo** [reply]
what if you put the types package in the root of your monorepo?

**Reply by nandorojo** [reply]
i have RN 0.71.3 and it’s working fine. same package for types

**Reply by midrizi** [reply]
> what if you put the types package in the root of your monorepo?

I did try `nohoist` but doesn't work. And yeah just copied the theme to a new react project and it works, so idk where the issue is.

```
yarn why v1.22.19
[1/4] Why do we have the module "@types/react-native"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@root/native#@types/react-native@0.71.5"
info Reasons this module exists
   - "_project_#@root#native" depends on it
   - in the nohoist list ["/_project_/@root/native/@types/react-native"]
info Disk size without dependencies: "804KB"
info Disk size with unique dependencies: "1000KB"
info Disk size with transitive dependencies: "2.16MB"
info Number of shared dependencies: 2
Done in 1.61s.
```

**Reply by midrizi** [reply]
I believe it's definitely some other package messing around with either React Types or with TypeScript, I will send u the demo project. 

This is happening in the Ignite boilerplate https://github.com/infinitered/ignite

**Reply by midrizi** [reply]
@nandorojo 

I did find the issue and it has to do with TypeScript config `strictNullChecks`

Same example as yours except with `strictNullChecks=false`:
https://www.typescriptlang.org/play?strictNullChecks=false#code/JYWwDg9gTgLgBAbzgZQB4Bo4gIYGsCmAKgBb4j5wC+cAZlBCHAOQAmUwYAzgJ5MBQfAMYQAdp3gxS5OAF4seIlPwAKBHzga4wgDbROALkTrNJgCQBzKPm6dB2bfkNqTLzQCNoLfFENM32gFd8flcNSmMwvkoASgEYbjAKEjIKOXjEiBo4SRSBL0FtbCssCBYAh2Y2Dh4mIxNgERhvGmxBCgARdi5uAGEA8QZk6XxUJpEWTjghigRw8L4HeE5UQzRZRHnFuGWAJlXUOWd3c18LKxs7BwA6DygvKCZ0CLhSYHNiGENlaNkAPjgAAxPExXUHLKJAA

<image width="300" src="https://user-images.githubusercontent.com/56638730/232029571-0a3d4688-a21a-47e3-8f6e-19cfa5c91cd5.png" />

**Reply by nandorojo** [reply]
Ohh interesting...I never let that stay false. Could you enable it?

**Reply by midrizi** [reply]
It does get rid of that error, but not all styles are recognized

![image](https://user-images.githubusercontent.com/56638730/232054738-2f9de1ad-794a-4fa6-9658-ce8305ff8c14.png)


Same with new react-native project here: https://github.com/midrizi/dripsy-demo 
in the `Divider.tsx` file

**Reply by midrizi** [reply]
@nandorojo 

You can try `dripsy/examples/example` as well, it's the same issue


<img width="700" src="https://user-images.githubusercontent.com/56638730/232250656-f6944511-bc57-45d2-9890-ac8762dc0fbc.png" />

**Reply by nandorojo** [reply]
i'll check but i think mine was working...are you sure your vscode is set to use TS 5?

**Reply by nandorojo** [reply]
command shift p → typescript version → workspace version

**Reply by midrizi** [reply]
@nandorojo Same issue

![image](https://user-images.githubusercontent.com/56638730/232308129-19c28e88-798c-4d6c-928e-f15cf2365a96.png)

**Reply by midrizi** [reply]
@nandorojo Seems like the following are not the same:

If is use it like so, it will not break the types:
```
types: {
  onlyAllowThemeValues: {
    borderStyles: 'always',
    borders: 'always',
    colors: 'always',
    space: 'always',
  },
  reactNativeTypesOnly: true,
}
```

But if you use it like this it does:
```
types: {
  onlyAllowThemeValues: 'always',
  reactNativeTypesOnly: true,
}
```

Can you check if this is the same on your end?

**Reply by nandorojo** [reply]
interesting. does this reproduce in the TS playground i sent before?

**Reply by midrizi** [reply]
@nandorojo 

Yes, I added all the values and doesn't throw an error, only when you use `onlyAllowThemeValues: 'always'`

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZQB4Bo4gIYGsCmAKgBb4j5wC+cAZlBCHAOQAmUwYAzgJ5MBQfAMYQAdp3gxS5OAF4seIlPwAKBHzga4MbmHycAXInWaTAelNxRAG24BBK1YgB3EmXwA1bFYCuew2pMzCxMvfFgDZi8nbG5OJnRjQLhzQK9gbE4-SKto2PjEwJSTACNsFgBzLKYomLiEpI0izWLoFjCI6pza-Ibk4ObWsORtKyqavPqGpo0WqDaoAHVgFkkO8bqCoMDiqHw8SGARGDWuic3Nabhi7xgYURPcjd7LwWw5h+7JpJeIRygPs7PfoaYR-YbcUYAp5TYFwQTecQMABiomOhk6jx6MMCNFRUKx31huKOyGAAC8xqdoYScaiFvhgOViGjspivoUidAQPj2VsTOV2CweedGrDQNhKsKgYFDsAYOkrABhX7QACyEDaADlsOR0esCRzAlYYhAbnqqQa+ZpRrchmBsIJDuUpdiTFZDnsoABxKBlYD4I4umluj0ACQZTJZGM+Ir6RsOuCDhpM5BY6QAir52JS2bHLuROJwJTmY9KTBB7Y75SXAa7NGB2EdlX91Vqdfhzbmy5pfWngJ3S3WNJxiGVnEmrcPyTXqcnNJxKx3WYPg-ORjPLRdYTB8KgYAPa6uNDu98hRyxxwfZ5OtLuYODIVfN6LAjBfWI5cB7k-eVvAgj8AAIQ1bgHw3X8XxMADgLmMJgNQH881hADm2gZBBCUVV8DTbBMzCbhEO7DQUIRO4QAABXoXRYH9Cc-ygzIABkIFeKxhmgYtCKHOAyQASRENNBHApDAmEEQaEZLijzgX0nFQ-4pMNSgEljaw7AcZxXHITwfA3AplIEExQWgCIAgaAASAV8FiVilzM3oZkGKB0WKXT+F6Sh9L4SgAEoBG0XQ4C0ig5AC-AIBoLQlAENpBGNXYsA1bxRmYNgOB4JgjBMQ4dygGgHQoAARdguG4RVSIYYK4DvAMhSCpRECobyBBtOBOAQlBUFkRBPL4Vr2oAJkMNBuvsuA0nKEReJ3bl0SEo4wgNYpynRSzdhs0IADpZnmA1SEZZlDGUHzZAAPjgAAGdlNpu9rvKAA
